include/cuco/detail/dynamic_map.inl

```diff
@@ -237,6 +237,52 @@ void dynamic_map<Key, Value, Scope, Allocator>::find(InputIt first,
   CUCO_CUDA_TRY(cudaDeviceSynchronize());
 }
 
+template <typename Key, typename Value, cuda::thread_scope Scope, typename Allocator>
+template <typename KeyOut, typename ValueOut>
+std::pair<KeyOut, ValueOut> dynamic_map<Key, Value, Scope, Allocator>::retrieve_all(
+  KeyOut keys_out, ValueOut values_out, cudaStream_t stream) const
+{
+  auto constexpr block_size = 128;
+  auto constexpr stride     = 1;
+
+  auto const capacity = get_capacity();
+  auto grid_size      = (capacity + stride * block_size - 1) / (stride * block_size);
+
+  std::vector<size_t> submap_cap_prefix(submaps_.size());
+  std::inclusive_scan(
+    submaps_.begin(),
+    submaps_.end(),
+    submap_cap_prefix.begin(),
+    [](auto const& sum, auto const& submap) { return sum + submap->get_capacity(); },
+    (size_t)0);
+  thrust::device_vector<size_t> submap_cap_prefix_d(submap_cap_prefix);
+
+  using temp_allocator_type =
+    typename std::allocator_traits<Allocator>::template rebind_alloc<char>;
+  auto temp_allocator = temp_allocator_type{alloc_};
+  auto d_num_out =
+    reinterpret_cast<unsigned long long*>(std::allocator_traits<temp_allocator_type>::allocate(
+      temp_allocator, sizeof(unsigned long long)));
+  CUCO_CUDA_TRY(cudaMemsetAsync(d_num_out, 0, sizeof(unsigned long long), stream));
```

Comment:
we have a [counter_storage](https://github.com/NVIDIA/cuCollections/blob/3e04706debe7ce8ef9f42f0ded26b62d6dfd1be1/include/cuco/detail/storage/counter_storage.cuh) utility dedicated to help in this situation, e.g. https://github.com/NVIDIA/cuCollections/blob/3e04706debe7ce8ef9f42f0ded26b62d6dfd1be1/include/cuco/detail/open_addressing/open_addressing_impl.cuh#L340-L342

Reply:
Yeah, using detail::counter_storage would be much more elegant, but I found it may not be consistent with current interface. The stream is a `cuda::stream_ref` in counter_storage, and `cudaStream_t` in cuco::dynamic_map.
Perhaps we can use `counter_storage` in `cuco::experimental::dynamic_map`  (if we add a `retrieve_all` method there)?

Reply:
Good point. Let me resolve this.